### Natural Computing Assignment 3 
Simge Ekiz(s4706757), Luca Parola(s1009497), Katrin Bujari(s1005213)

### Question 1


Consider an illustrative example of a PSO system composed of three particles. Consider the following update rule for each particle i and dimension d:
\begin{equation}
		v(i;d) = wv(i;d) + r_1({x}^*(i;d) - x(i;d)) + r_2({x}^*(d) - x(i;d)) 
\end{equation}


To facilitate calculation, we will ignore the fact that $r_1$ and $r_2$ are random numbers and fix them to $0.5$ for this exercise. The space of solutions is the two dimensional real valued space and the current state of the swarm is as follows

-  Position of particles: $x_1 = (5,5); x_2 = (8,3); x_3 = (6,7)$
-  Individual best positions: $x^*_1 = (5,5); x^*_2 = (7,3); x^*_3 = (5,6)$
-  Social best position: $x^* = (5,5)$
-  Velocities: $v_1 = (2,2); v_2 = (3,3); v_3 = (4,4)$


a-) What would be the next position of each particle after one iteration of the PSO algorithm with w = 2?



<div style="padding-left: 20px;">
$w = 2$
$r_1 = 0.5$
$r_2 = 0.5$
<br><br>
<b> According to formula; </b><br>
  <span style="padding-left: 15px;" >  $v(i;d) = wv(i;d) + r_1(x^∗(i;d) - x(i;d)) + r2(x^∗(d) - x(i;d))$
  </span>
<br><br>
<b> Velocity of particle 1; </b><br> 
<span style="padding-left: 15px;" >
    $v(1;xaxis) = 2*2 + 0.5*(5-5) + 0.5*(5-5) = 4$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(1;yaxis) = 2*2 + 0.5*(5-5) + 0.5*(5-5) = 4$ <br> 
</span>
<span style="padding-left: 15px;" >
    ${v}_1^2 = (4, 4)$ 
</span>
<br><br>
<b> Velocity of particle 2;</b> <br>
<span style="padding-left: 15px;" >
    $v(2;xaxis) = 2*3 + 0.5*(7-8) + 0.5*(5-8) = 4$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(2;yaxis) = 2*3 + 0.5*(3-3) + 0.5*(5-3) = 7$ <br>
</span>
<span style="padding-left: 15px;" >
    $v_2^2 = (4, 7)$ 
</span>
<br><br>
<b> Velocity of particle 3; </b><br>
<span style="padding-left: 15px;" >
    $v(3;xaxis) = 2*4 + 0.5*(5-6) + 0.5*(5-6) = 7$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(3;yaxis) = 2*4 + 0.5*(6-7) + 0.5*(5-7) = 6.5$ <br>
</span>
<span style="padding-left: 15px;" >
    $v_3^2 = (7, 6.5)$ 
</span>
<br><br>
<b> Updating positions : </b> <br>
<span style="padding-left: 15px;" >
    $x_i^{t+1} = x_i^{t} + v_i^{t+1}$
</span><br><br>
<b> Position of particle 1; </b> <br>
<span style="padding-left: 15px;" >
$x_1^2 = (5, 5) + v_1^2 = (5, 5) + (4, 4) = (9, 9) $
</span>
<br><br>
<b> Position of particle 2; </b><br>
<span style="padding-left: 15px; " >
$x_2^2 = (8,3) + v_2^2 = (8, 3) + (4, 7) = (12, 10) $ 
</span> <br><br>
<b>Position of particle 3; </b> <br>
<span style="padding-left: 15px;" >
$x_3^2 = (6,7) + v_3^2 = (6, 7) + (7, 6.5) = (13, 13.5) $ 
</span>
</div>

b-) And using w = 0.1?

<div style="padding: 20px;">
$w=0.1$
$r_1 = 0.5$
$r_2 = 0.5$
<br><br>
<b>Velocity of particle 1;</b> <br> 
<span style="padding-left: 15px;" >
    $v(1;xaxis) = 0.1*2 + 0.5*(5-5) + 0.5*(5-5) = 0.2$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(1;yaxis) = 0.1*2 + 0.5*(5-5) + 0.5*(5-5) = 0.2$ <br> 
</span>
<br>
<b>Velocity of particle 2;</b> <br>
<span style="padding-left: 15px;" >
    $v(2;xaxis) = 0.1*3 + 0.5*(7-8) + 0.5*(5-8) = -1.7$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(2;yaxis) = 0.1*3 + 0.5*(3-3) + 0.5*(5-3) = 1.3$ <br>
</span>
<br>
<b>Velocity of particle 3;</b> <br>
<span style="padding-left: 15px;" >
    $v(3;xaxis) = 0.1*4 + 0.5*(5-6) + 0.5*(5-6) = -0.6$ <br>
</span>
<span style="padding-left: 15px;" >
    $v(3;yaxis) = 0.1*4 + 0.5*(6-7) + 0.5*(5-7) = -1.1$ <br>
</span>
<br><br>

<b>Position of particle 1;</b> <br>
<span style="padding-left: 15px;" >
$x_1^2 = (5, 5) + (0.2, 0.2) = (5.2, 5.2) $
</span>
<br><br>
<b>Position of particle 2;</b> <br>
<span style="padding-left: 15px; " >
$x_2^2 = (8, 3) + (-1.7, 1.3) = (6.3, 4.3) $ 
</span> <br><br>
<b>Position of particle 3; </b> <br>
<span style="padding-left: 15px;" >
$x_3^2 = (6, 7) + (-0.6, -1.1) = (5.4, 5.9) $ 
</span>
</div>

c-) Explain what is the effect of the parameter w.
> In PSO, for each particle, velocity is calculated according to its previous velocity, the particle location which has the best fitness so far and any particles location which has the best fitness so far. W carries a factor that how much the previous velocity contribute to the next velocity. small w facilitates local search, it exploits the current position as much as it can. Large w facilitates global search, it aims to explore solution space.


d-) Give an advantage and a disadvantage of a high value of w.
> An advantage of high value w is that it allows us to explore more. Particles move faster and more solution space could be searced and this increase the possibility of finding a better solution.
A disadvantage of of high value w is that;
it does less expoitation and gets less sensitive to new information in the influence from the current particle. Thus, it might skip over an optimum.





### Question 2

Consider a particle swarm consisting of a single member. How would it perform in a
trivial task such as the minimization of $f(x) = x^2$ when $w < 1$?

> Given that w<1 and $f(x) = x^2$ it would find the optimum solution. Finding a solution depends on the objective function since f(x) establishes a convex solution space which has less complexity of the search space. For instance, if the particle starts with the velocity which drawns the particle away from the optimum, the personal best affects the particle and after some time particle will be moving to an optimum point. When velocity is in the same direction with the optimum, personal best loses its influence. Then, velocity starts to decrease. If the particle reaches to optimum before the velocity is zero, algorithm finds the optimum point. Particle will find the optimum for this trivial problem, because when w < 1, particles would be moving around the optimum. Then, after some steps, velocity would be zero at the optimum.


### Question 3

Implement the PSO algorithm for clustering described in “Van der Merwe, D. W., and Andries Petrus Engelbrecht. ”Data clustering using particle swarm optimization.” Evolutionary Computation, 2003. CEC’03. The 2003 Congress on. Vol. 1. IEEE, 2003.” (see also swarm intelligence slides). Implement the k-means clustering.
Apply and compare the performance of the two algorithms in terms of quantization error on Artificial dataset 1 and Iris dataset (the latter available at UCI ML repository, see https://archive.ics.uci.edu/ml/datasets/iris). In both algorithms, use the true number of clusters as value of the parameter for setting the number of clusters.

In [1]:
from sklearn import datasets
from sklearn.cluster import KMeans
import pandas as pd
import numpy as np
import scipy

We build the Artificial dataset defined as the equation 9 in Van der Merwe, D. W., and
Andries Petrus Engelbrecht. "Data clustering using particle swarm optimization."

In [2]:
z1=np.random.uniform(-1,1,400)
z2=np.random.uniform(-1,1,400)

In [3]:
artificial_df=pd.DataFrame({"Z1" : z1,"Z2" : z2})

In [4]:
kmeans1 = KMeans(n_clusters=2, random_state=0).fit(artificial_df)

We predict the class of each datapoint

In [5]:
prediction_artificial_df=kmeans1.predict(artificial_df)

Now we can compute the quantization error for the artificial dataset

### quantization error

$$J_{e}\,=\,{{\sum_{j=1}^{N_{c}}[\sum_{\forall {\bf z}_{{\rm p}}\in C_{ij}} d({\bf z}_{p},{\bf m}_{j})/\vert C_{ij}\vert]}\over{N_{c}}}$$


$$d ( {{\bf z}_{p}, m_{j}})\,=\,\sqrt{\sum_{k=1}^{N_{j}}(z_{pk}-m_{jk})^{2}}$$



$N_{c}$ denotes the number of cluster centroids (as provided by the user), i.e. the number of dusters to be formed

${\bf z}_{p}$ denotes the p-th data vector

$C_{j}$ is the subset of data vectors that form cluster j.

$m_{j}$ is the centroid 

$d_{{\bf z}_{p}, m_{j}}$ is the Euclidean distance 


In [6]:
##we take the centroids:
centroid1_A_df=kmeans1.cluster_centers_[0]
centroid2_A_df=kmeans1.cluster_centers_[1]

In [7]:
## we assign each data point to the predicted cluster 
cluster1_A=[]
cluster2_A=[]
for i in range(0,len(artificial_df)):
    if prediction_artificial_df[i]==0:
        
        cluster1_A.append(artificial_df.loc[i])
        
    if prediction_artificial_df[i]==1:
        cluster2_A.append(artificial_df.loc[i])

In [8]:
def quantization_error2(cluster1,cluster2,centroid1,centroid2):
            
    distance1=0
    distance2=0


    for i in range(0,len(cluster1)):

        dist1=scipy.spatial.distance.euclidean(np.asarray(cluster1[i]), centroid1)
        distance1=distance1+dist1
    
    if len(cluster1)>0:
        distance1=distance1/len(cluster1)
    else:
        distance1=0

    for i in range(0,len(cluster2)):

        dist2=scipy.spatial.distance.euclidean(np.asarray(cluster2[i]), centroid2)
        distance2=distance2+dist2
    
    if len(cluster2)>0:
        distance2=distance2/len(cluster2)
    else:
        distance2=0


    QE=(distance1+distance2)/2
    ##print("Quantization error: ",QE)
    return QE

In [9]:
print ("The quantization error for the KMEANS algorithm for the artificial dataset is  ", quantization_error2(cluster1_A,cluster2_A,centroid1_A_df,centroid2_A_df))

The quantization error for the KMEANS algorithm for the artificial dataset is   0.5686478747859159


Now we can do the same with the iris dataset

In [10]:
iris = datasets.load_iris()
y=iris.target
iris_df = pd.DataFrame(iris.data, columns=iris.feature_names)

In [11]:
kmeans2 = KMeans(n_clusters=3, random_state=0).fit(iris_df)

In [12]:
prediction_iris_df=kmeans2.predict(iris_df)

In [13]:
##we take the centroids:
centroid1_iris=kmeans2.cluster_centers_[0]
centroid2_iris=kmeans2.cluster_centers_[1]
centroid3_iris=kmeans2.cluster_centers_[2]


In [14]:
## we assign each data point to the predicted cluster 
cluster1_iris=[]
cluster2_iris=[]
cluster3_iris=[]
for i in range(0,len(iris_df)):
    if prediction_iris_df[i]==0:
        cluster1_iris.append(iris_df.loc[i])
    if prediction_iris_df[i]==1:
        cluster2_iris.append(iris_df.loc[i])
    if prediction_iris_df[i]==2:
        cluster3_iris.append(iris_df.loc[i])

We define our qunatization error function for 3 cluster

In [15]:
def quantization_error3(cluster1,cluster2,cluster3,centroid1,centroid2,centroid3):
            
    distance1=0
    distance2=0
    distance3=0
    count=0
    for i in range(0,len(cluster1)):

        dist1=scipy.spatial.distance.euclidean(np.asarray(cluster1[i]), centroid1)
        distance1=distance1+dist1
    
    if len(cluster1)>0:
        distance1=distance1/len(cluster1)
    else:
        distance1=0
    
   
    

    for i in range(0,len(cluster2)):

        dist2=scipy.spatial.distance.euclidean(np.asarray(cluster2[i]), centroid2)
        distance2=distance2+dist2
    
    if len(cluster2)>0:
        distance2=distance2/len(cluster2)
    else:
        distance2=0
    
  
    
    
    for i in range(0,len(cluster3)):

        dist3=scipy.spatial.distance.euclidean(np.asarray(cluster3[i]), centroid3)
        distance3=distance3+dist3
    
    if len(cluster3)>0:
        distance3=distance3/len(cluster3)
    else:
        distance3=0


    
    QE=(distance1+distance2+distance3)/3
    return QE

In [16]:
print ("The quantization error for the KMEANS algorithm for the iris dataset is  ", quantization_error3(cluster1_iris,cluster2_iris,cluster3_iris,centroid1_iris,centroid2_iris,centroid3_iris))

The quantization error for the KMEANS algorithm for the iris dataset is   0.6473743892616001


Now we can build the PSO algorithm and use it with the Artificial DATaset

In [17]:
#we initialize 10 particles with 2 centroid each to use in the Artificial dataset
particles_list=[]
for i in range(0,10):
    particles=np.random.uniform(-1,1,4).reshape(2,2)
    particles_list.append(particles)

In [18]:
##we set our initial values

personal_best_error=np.ones(10)
global_best_position=[]
global_best_error=1
personal_best_position=np.zeros(40).reshape(10,2,2)

alpha=1.49618
w=0.7298
velocity=np.zeros(4).reshape(2,2) 

##for each iteration
for iteration in range(0,10):
    
    quantization_error=[] 
    count=0
    
    ##for each particles 
    for i in range(0,len(particles_list)):

        
        current_particle=particles_list[i]
        cluster1=[]
        cluster2=[]
        
       ##for each data point in the Artificial dataset
        for a in range(0,len(artificial_df)):
            
            
            ##we measure the euclidean distance from each centroid
            dist1=scipy.spatial.distance.euclidean(np.asarray(artificial_df.loc[a]), current_particle[0])
            dist2=scipy.spatial.distance.euclidean(np.asarray(artificial_df.loc[a]), current_particle[1])

            ##we assign each datapoint to a cluster based on the closest centroid
            if dist1<dist2:
                cluster1.append(artificial_df.loc[a])


            elif dist1>dist2:
                cluster2.append(artificial_df.loc[a])


        
        ##Once we have our cluster we compute the quantization error and we store it
        ## we will have a list with all our quantization error
        quantization_error.append(quantization_error2(cluster1,cluster2,current_particle[0],current_particle[1]))

   
    
    ##Then for each quantization error in the list
    for z in range(0,len(quantization_error)):
        count=count+1    
        
        ##if the quantization error of the current iteration is smaller than the personal best , we store it as new personal best
        if quantization_error[z]<personal_best_error[z]:
            personal_best_error[z]=quantization_error[z]
            ##and we store the corresponding particle as new best position
            personal_best_position[z]=particles_list[z]

        ##we also update the best global error if we find a new best 
        if personal_best_error[z]<global_best_error:
            global_best_error=personal_best_error[z]
            global_best_position=particles_list[z]
            
    ###print ("Particle list BEFORE, ", particles_list)
    
    
    
    
    ## then for each particle we update the position of the centroids based on the formula
    
    for i in range(0,len(particles_list)):
        
        ##I've disassembled the formula for clarity.

        

        ##we set the random r
        r1=np.random.uniform(0,1,1)
        r2=np.random.uniform(0,1,1)

        ##we compute the first term of the formula
        first_term=np.multiply(w,velocity)
        
        ##we compute the two multiplication of the alpha and the r
        alphaR1=alpha*r1
        alphaR2=alpha*r2
        
        #we copmute the 2 subtractions
        first_subtraction=np.subtract(personal_best_position[i],particles_list[i])
        second_subtraction=np.subtract(global_best_position,particles_list[i])

        #we compute the second term
        second_term=np.multiply(alphaR1,first_subtraction)
        
        #we compute the second term
        third_term=np.multiply(alphaR2,second_subtraction)
        
        ##we compute the velocity
        velocity=np.add(first_term,second_term)
        velocity=np.add(velocity,third_term)

        #we update the position of the centroids of the particles
        particles_list[i]=np.add(particles_list[i],velocity)
        
    ##print ("global best error is: ", global_best_error)
    ##print ("global best position: ", global_best_position)
    ##print ("personal best error: ", personal_best_error)
    ##print ("Particle list AFTER", particles_list)
    ##print("personal best position  ",personal_best_position )

    
print("After ",iteration," iteration , the global best error is: ", global_best_error, "best  swarn has positions: ", global_best_position)

After  9  iteration , the global best error is:  0.37917153606964155 best  swarn has positions:  [[ 0.64776438 -4.26549274]
 [ 0.06790201 -0.10070954]]


Now we can di the same thing for the irisi dataset. This time as we have 3 classes and 4 dimensions we set each particle to have 3 centroids based on 4 dmensions

We take the max and min of each columns in the iris dataset, so we can set limits to our randomly chosen centroids

In [19]:
max_sepal_length=max(iris_df.iloc[:,0])
min_sepal_length=min(iris_df.iloc[:,0])

max_sepal_width=max(iris_df.iloc[:,1])
min_sepal_width=min(iris_df.iloc[:,1])

max_petal_length=max(iris_df.iloc[:,2])
min_petal_length=min(iris_df.iloc[:,2])

max_petal_width=max(iris_df.iloc[:,3])
min_petal_width=min(iris_df.iloc[:,3])

In [20]:
#we initialize 10 particles with 3 centroid each and 4 dimensions to use in the iris dataset
particles_list_iris=[]

for i in range(0,10):
    particle=[]
    
    for i in range(0,3):
        centroid=[np.random.uniform(min_sepal_length,max_sepal_length,1),np.random.uniform(min_sepal_width,max_sepal_width,1),np.random.uniform(min_petal_width,max_petal_width,1),np.random.uniform(min_petal_width,max_petal_width,1)]
        particle.append(centroid)
    
    particles_list_iris.append(particle)
    
particles_list_iris=np.asarray(particles_list_iris).reshape(10,3,4)

Now we can run the algorithm

In [21]:
##we set our initial values

personal_best_error=np.full((1, 10), 50).astype(float)
personal_best_error=personal_best_error[0][:]
global_best_position=[]
global_best_error=100  ##we set this to 100 so we are sure that the algorithm will find a smaller value
personal_best_position=np.zeros(120).reshape(10,3,4)

alpha=1.49618
w=0.7298
velocity=np.zeros(12).reshape(3,4) 

##for each iteration
for iteration in range(0,20):
    
    quantization_error=[] 
    count=0
    
    ##for each particles 
    for i in range(0,len(particles_list_iris)):

        
        current_particle=particles_list_iris[i]
        cluster1=[]
        cluster2=[]
        cluster3=[]
        
       ##for each data point in the Artificial dataset
        for a in range(0,len(iris_df)):
            
            
            ##we measure the euclidean distance from each centroid
            dist1=scipy.spatial.distance.euclidean(np.asarray(iris_df.loc[a]), current_particle[0])
            dist2=scipy.spatial.distance.euclidean(np.asarray(iris_df.loc[a]), current_particle[1])
            dist3=scipy.spatial.distance.euclidean(np.asarray(iris_df.loc[a]), current_particle[2])
            
            
            ##we assign each datapoint to a cluster based on the closest centroid
            if dist1<dist2 and dist1<dist3:
                cluster1.append(iris_df.loc[a])


            elif dist2<dist3 and dist2<dist3:
                cluster2.append(iris_df.loc[a])
                
            elif dist3<dist1 and dist3<dist1:
                cluster3.append(iris_df.loc[a])


        
        ##Once we have our cluster we compute the quantization error and we store it
        ## we will have a list with all our quantization error
        quantization_error.append(quantization_error3(cluster1,cluster2,cluster3,current_particle[0],current_particle[1],current_particle[2]))

   
    
    ##Then for each quantization error in the list
    for z in range(0,len(quantization_error)):
        count=count+1    
        
        ##if the quantization error of the current iteration is smaller than the personal best , we store it as new personal best
        if quantization_error[z]<personal_best_error[z]:
            personal_best_error[z]=quantization_error[z]
            ##and we store the corresponding particle as new best position
            personal_best_position[z]=particles_list_iris[z]

        ##we also update the best global error if we find a new best 
        if personal_best_error[z]<global_best_error:
            global_best_error=personal_best_error[z]
            global_best_position=particles_list_iris[z]
            
    ###print ("Particle list BEFORE, ", particles_list)
    
    
    
    
    ## then for each particle we update the position of the centroids based on the formula
    
    for i in range(0,len(particles_list_iris)):
        
        ##I've disassembled the formula for clarity.

        

        ##we set the random r
        r1=np.random.uniform(0,1,1)
        r2=np.random.uniform(0,1,1)

        ##we compute the first term of the formula
        first_term=np.multiply(w,velocity)
        
        ##we compute the two multiplication of the alpha and the r
        alphaR1=alpha*r1
        alphaR2=alpha*r2
        
        #we copmute the 2 subtractions
        first_subtraction=np.subtract(personal_best_position[i],particles_list_iris[i])
        second_subtraction=np.subtract(global_best_position,particles_list_iris[i])

        #we compute the second term
        second_term=np.multiply(alphaR1,first_subtraction)
        
        #we compute the second term
        third_term=np.multiply(alphaR2,second_subtraction)
        
        ##we compute the velocity
        velocity=np.add(first_term,second_term)
        velocity=np.add(velocity,third_term)

        #we update the position of the centroids of the particles
        particles_list_iris[i]=np.add(particles_list_iris[i],velocity)
        
    ##print ("global best error is: ", global_best_error)
    ##print ("global best position: ", global_best_position)
    ##print ("personal best error: ", personal_best_error)
    ##print ("Particle list AFTER", particles_list)
    ##print("personal best position  ",personal_best_position )

    
print("After ",iteration," iteration , the global best error is: ", global_best_error, "best  swarn has positions: ", global_best_position)

After  19  iteration , the global best error is:  0.8565857081242823 best  swarn has positions:  [[-3.32146349e+08 -9.38001868e+06  5.87334275e+07  7.11057973e+07]
 [-4.09695375e+07  1.79198894e+07 -2.72622289e+07 -1.08861167e+08]
 [ 3.61659108e+08  6.46734422e+07  1.74208476e+08 -2.97542908e+07]]


If we compare the results with the results described in the paper, we can notice that our PSO and KMEANS algorithm performs somehow similarly to the results of the paper. In fact we have:

- Quantization error for the KMEans algorithm for the articifial dataset:   0.5778295493707196
- Quantization error for the KMEans algorithm for the iris dataset:   0.6473743892616001

- Quantization error for the PSO algorithm for the articifial dataset:  0.4570523934830559 
- Quantization error for the PSO algorithm for the articifial dataset::   0.8999702242159057 

#### Question 4

The figure shows an example from the ACO book by Dorigo and Stuetzle. What results do you expect for an ant colony algorithm that does not use tabu lists (except for inhibition of immediate return to the previous node)?

<img src="ACO.png" alt="Drawing" style="width: 30%;"/>

> ACO is inspired from the foraging behaviour of ants and imitates how they find the closest food sources to their nests. They communicate with pheromones they leave behind while moving. Likewise, they tend to choose paths marked by strong pheromone concentrations. When an ant finds a food source, it leaves some amount of pheromone that may depend on the quantity and quality of the food during the return trip. Thus, quality food sources closest to the nest will have the path with more pheromone. 
However in this case without a tabu list, it is unavoidable that ants make cycles. At the bottom of the picture above, there are a lot of connected paths and this will cause a lot of pheromone accumulation and will result with most likely loopy paths. For such situations converging the minumum path is not trivial.
To find the optimum solution an ant has to make number of correct desicions, otherwise ant generates suboptimal solutions.
There is a trade off between use of an easy but suboptimal path and searching the optimal path. 
In this case, easy path would be that the ants take the route through the upper nodes. There are two optimal paths which are in the lower part of the graph and harder to find.
We can avoid from cycles with high pheromone evaporation. However, this may end up with not generating optimum solutions. Since this problem is relatively simple and can be solved with a large number of ants. After a lot of iterations algorithm find the fastest route in the densely connected part of graph which is down-up-right-down-up or down-down-rigt-up-up.


#### Question 5

Assume that ants are allowed to lay pheromone on a path at every time step, so that the pheromone update rule is applied at each time step. Come up with a combination local/global updating scheme that encourages exploration and exploitation - consider which parameters influence this.

> Ants probabilisticaly choose the next node based on the amount of pheromone while it is searching for food and exploring the space.
For each ant we can keep a limited memory which is local pheromone. This memory allows them to record their trip to their destination. When they reach their destination they can construct a way back to home. If they've passed over the same node twice they can eliminate the path in between those repeated nodes. Then they will retrace the steps they constructed based on the memory(local pheromone). On the way back to the nest ant will ignore the pheromone trails, follow the path that is costructed.
When they are going back to the nest they will be leaving pheromones to the ground(global pheromones). They will not apply to the global pheromones on the food searching step because that can cause loops.
After each round the new pheromones are added to the old ones. For better exploitation, good ants should influence the new pheromones more than others.


#### Question 6

The goal of this exercise is to practice with an implementation of ACO, so please do not use other (possibly better) methods for solving this problem.

(a) Code an ACO to solve Sudoku’s. First, you need to think how to represent Sudoku: there are three conditions an optimal solution must fulfill, each line must contain all integers {1, 2, 3, 4, 5, 6, 7, 8, 9} once and only once, the same applies to columns and 3x3 subgrids. You can fix one of them so that you will prevent violations. The ACO will have 9x9x9 pheromone matrix, where you update pheromones for the values appearing in the best solutions inversely proportionally to the fitness value. You have to consider also how the old pheromones will fade.


In [22]:
import numpy as np
from collections import Counter
import math
import sys
import random

# inputfile = "sudoku/s10b.txt"
iteration_number = 10 
r = 0.05
ants = 5

In [23]:
def ACO_Algorithm(a, t, ants):
    sudoku_problem = a.copy()   
    bestfitness = 0             #initialize the best fitness
    bestsolution = []           # initialize the best solution
    for i in range (1, ants+1):             # create all ants
        fitness, solution = ant(sudoku_problem,t)  # retrieve the solution and fitness from the ant given the problem and pheramone matrix
        if fitness > bestfitness:           # compare fitness to best fitness so far 
            bestfitness = fitness           # save new fitness value if greater than previous
            bestsolution = solution.copy()  # copy the solutions
            if np.count_nonzero(solution) == 81:   # if there is no zero left then the answer is found 
                break                       
    return bestfitness, bestsolution        # return best solution and fitness for one iteration


In [24]:
def update_solution(ant_solution, digits, row_matrix, digit_submatrix, column_matrix):
    # for each position if there are obvious digits to fill, Fill the digit with that number
    for (i,j), p in np.ndenumerate(digits):
        # calculate index z for digit_submatrix
        if i < 3: 
            z = int((j/3)+1)
        elif 2 < i < 6: 
            z = int((j/3)+1)+3
        elif 5 < i: 
            z = int((j/3)+1)+6

        # calculate the list of digits that are taken for this position
        digit_list = list(row_matrix[i] + column_matrix[j] + digit_submatrix[z-1])

        # count how many are available to fill
        digit_possible = digit_list.count(0) 

        # if there are possible digits, save the number in the digits matrix
        if ant_solution[i][j] > 0: 
            digit_possible = 0
        digits[i][j]=digit_possible

        # if only one digit is possible, find it and put it in the solution
        if digit_possible == 1 and ant_solution[i][j] == 0:
            ant_solution[i][j]=digit_list.index(0)+1
            digits[i][j]=0
            break

    return ant_solution, digits

In [25]:
def update_memory(ant_solution, row_matrix, digit_submatrix, column_matrix):
    # Create a memory of taken numbers in submatrixes, rows and columns
    count = 0
    for x in [0,3,6]:
        for y in [0,3,6]:
            for (i2,j2),sd in np.ndenumerate(ant_solution[x:x+3,y:y+3]):
                if sd > 0:
                    digit_submatrix[count][sd-1]=1
            count += 1
    # for each row and column, check which numbers are already taken
    for (i,j), digit in np.ndenumerate(ant_solution):
        if digit > 0:
            row_matrix[i][digit-1] = 1
            column_matrix[j][digit-1] = 1

    return row_matrix, digit_submatrix, column_matrix

In [26]:

def ant(d,t):
    #initialize the antas solution set 
    ant_solution = d.copy()   

    current_fitness = 0
    row_matrix = np.zeros((9,9)) # initialize digit matrix in each row
    column_matrix = np.zeros((9,9)) # initialize digit matrix in each column
    digit_submatrix = np.zeros((9,9), dtype=np.int) # initialize the matrix for possible digits in each 3x3 submatrix
    digits = np.zeros((9,9)) # initialize the possible digits matrix
    while (True):
        
        row_matrix, digit_submatrix, column_matrix = update_memory(ant_solution, row_matrix, digit_submatrix, column_matrix)
        ant_solution, digits = update_solution(ant_solution, digits, row_matrix, digit_submatrix, column_matrix)
        
        # if there are no zeros in ant_solution answer is found, kill the ant and return the solution
        if np.count_nonzero(ant_solution) == 81: 
            return current_fitness, ant_solution
        
        #if no more steps are possible based on analytics
        if(np.count_nonzero(ant_solution)==current_fitness):
            w = np.zeros((9,9,9)) # initialize weight matrix
            wsum = 0 # sum of weights
            # for each weight, fill in the actual weight
            for (i,j,k), v in np.ndenumerate(w):
                if i < 3: 
                    z = int((j/3)+1)
                elif 2 < i < 6: 
                    z = int((j/3)+1)+3
                elif 5 < i: 
                    z = int((j/3)+1)+6
                
                if digits[i][j] > 0:
                    # weight equation: pheromone * 10-number of possible places for the digit * 10-number of possible digits for the position
                    w[i][j][k] = t[i][j][k]*(10-list(digit_submatrix[z-1]).count(0))*(10-digits[i][j])
                    # sum up the weights
                    wsum += w[i][j][k]
            
            # in case there are no more options left return
            if wsum == 0: 
                return current_fitness, ant_solution
            
            # calculate a random value between 0 and the sum of weights and normalize it
            next_selection = random.randint(0,int(wsum))/wsum
            # iterate through possible digits until the random value has been reached, effectively a weighted roulette wheel
            psum = 0
            for (i,j,k), v in np.ndenumerate(w):
                psum += w[i][j][k]/wsum
                # if this digit is the chosen one.. 
                if psum > next_selection:
                    # generate z index for 3x3
                    if i < 3: 
                        z = int((j/3)+1)
                    if 2 < i < 6: 
                        z = int((j/3)+1)+3
                    if 5 < i: 
                        z = int((j/3)+1)+6
                    # check for validity and add to solution
                    if row_matrix[i][k] == 0 and column_matrix[j][k] == 0 and digit_submatrix[z-1][k] == 0:
                        ant_solution[i][j] = k+1
                    else:
                        return current_fitness, ant_solution
                    
                    break
        else:
            current_fitness = np.count_nonzero(ant_solution)          
    return ant_solution, current_fitness

(b) Consider the benchmark Sudoku problems are available at http://lipas.uwasa.fi/~timan/sudoku/. Run your ACO on the following benchmark instances s10a.txt, s10b.txt, s11a.txt, s11b.txt. Report and discuss the results.

In [27]:
#open inputfile and take sudoku into problem_matrix
inputfiles = ["sudoku/s10a.txt", "sudoku/s10b.txt", "sudoku/s11a.txt","sudoku/s11b.txt"]
for inputfile in inputfiles:
    print("\nfor the problem at " + inputfile + "\n")
    # Create the matrixes;
    problem_matrix = np.zeros((9,9), dtype=np.int)  # initialize problem matrix    
    global_pheromone= np.ones((9,9,9))*1000           # initialize pheromene matrix
    globalbestfitness = 0                           # initialize global best fitness
    globalbestsolution = []                         # initialize global best solution matrix

    with open(inputfile, "r") as f:
        for i,line in enumerate(f):
            digitlist = line.split()
            if not digitlist == []:
                for j,digitstr in enumerate(digitlist):
                    problem_matrix[i][j] = int(digitstr)


    for iteration in range (1,iteration_number):
        # the best solution & fitness for iteration
        bestfitnessofiteration, bestsolutionofiteration = ACO_Algorithm(problem_matrix,global_pheromone,ants)
        # if it is better than the global best, copy it
        if(bestfitnessofiteration>globalbestfitness):
            globalbestfitness = bestfitnessofiteration
            globalbestsolution = bestsolutionofiteration.copy()
        # if it is solved, stop iteration
        if(np.count_nonzero(globalbestsolution) == 81):
            break
        # equation for new pheromone to be added
        add_pheromone = (bestfitnessofiteration/81)
        # pheromone decay according to rate r
        for (i,j,k), phe in np.ndenumerate(global_pheromone):
            global_pheromone[i][j][k] = (1-r)*global_pheromone[i][j][k]
        # add new pheromone for best solution set
        for (i,j), k in np.ndenumerate(bestsolutionofiteration):
            if int(k) > 0:
                # pheromone update equation, the constant can be modified to change behavior
                 global_pheromone[i][j][int(k-1)] += 1000*add_pheromone
    print(globalbestsolution)
    print('The problem is solved at iteration ' + str(iteration))


for the problem at sudoku/s10a.txt

[[9 7 2 8 6 3 5 4 1]
 [6 1 8 7 4 5 9 2 3]
 [4 5 3 2 9 1 6 8 7]
 [5 4 9 1 2 8 7 3 6]
 [8 2 1 6 3 7 4 5 9]
 [7 3 6 4 5 9 2 1 8]
 [2 9 5 3 8 6 1 7 4]
 [1 8 4 9 7 2 3 6 5]
 [3 6 7 5 1 4 8 9 2]]
The problem is solved at iteration 1

for the problem at sudoku/s10b.txt

[[8 5 2 7 1 6 9 4 3]
 [1 9 7 8 4 3 6 5 2]
 [4 6 3 9 2 5 1 8 7]
 [2 7 8 6 3 4 5 9 1]
 [6 4 5 1 7 9 3 2 8]
 [9 3 1 5 8 2 4 7 6]
 [7 8 6 4 9 1 2 3 5]
 [3 1 4 2 5 8 7 6 9]
 [5 2 9 3 6 7 8 1 4]]
The problem is solved at iteration 1

for the problem at sudoku/s11a.txt

[[3 0 5 0 7 0 2 0 0]
 [2 7 9 0 5 0 1 8 0]
 [8 0 1 0 2 0 5 0 7]
 [1 9 7 3 4 6 8 5 2]
 [4 5 2 7 1 8 3 9 6]
 [6 8 3 5 9 2 7 0 0]
 [7 0 8 0 6 0 9 0 5]
 [5 1 6 9 3 7 4 2 8]
 [9 0 4 0 8 0 6 0 0]]
The problem is solved at iteration 9

for the problem at sudoku/s11b.txt

[[1 0 0 7 6 9 2 8 5]
 [7 5 8 2 1 3 6 4 9]
 [9 6 2 4 8 5 1 7 3]
 [5 0 0 8 0 0 0 1 6]
 [2 0 0 0 0 0 0 0 4]
 [8 9 0 0 0 6 0 0 7]
 [6 1 9 3 7 8 4 5 2]
 [3 2 0 9 5 4 7 6 8]
 [4

s10a and s10b are solved in one iteration because they can be solved only using analytical methods. However for s11a and s11b and iteration number to solve them is between 1 and 10.
We choose 0.05 evaporation rate because we want pheromones to decay quickly but not quickly. Also, we use 5 ants 